File: cpp/tests/unit-tests/JoynrMessageFactoryTest.cpp
/*
 * #%L
 * %%
 * Copyright (C) 2017 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
#include <chrono>
#include <cstdint>
#include <string>

#include <gtest/gtest.h>

#include "joynr/JoynrMessageFactory.h"
#include "joynr/Request.h"
#include "joynr/Reply.h"
#include "joynr/MessagingQos.h"
#include "joynr/SubscriptionPublication.h"
#include "joynr/SubscriptionRequest.h"
#include "joynr/MulticastSubscriptionRequest.h"
#include "joynr/MulticastPublication.h"
#include "joynr/SubscriptionReply.h"
#include "joynr/SubscriptionStop.h"
#include "joynr/MulticastSubscriptionQos.h"
#include "joynr/OnChangeSubscriptionQos.h"

using namespace joynr;

class JoynrMessageFactoryTest : public ::testing::Test
{
public:
    JoynrMessageFactoryTest()
            : messageFactory(0),
              senderID(),
              receiverID(),
              requestReplyID(),
              qos(),
              request(),
              oneWayRequest(),
              reply(),
              subscriptionPublication()
    {
    }

    void SetUp()
    {
        senderID = std::string("senderId");
        receiverID = std::string("receiverID");
        requestReplyID = "requestReplyID";
        qos = MessagingQos(456000);
        qos.putCustomMessageHeader("test-header", "test-header-value");
        request.setMethodName("methodName");
        request.setRequestReplyId(requestReplyID);
        request.setParams(42, "value");
        request.setParamDatatypes({"java.lang.Integer","java.lang.String"});
        oneWayRequest.setMethodName("methodName");
        oneWayRequest.setParams(42, "value");
        oneWayRequest.setParamDatatypes({"java.lang.Integer","java.lang.String"});
        reply.setRequestReplyId(requestReplyID);
        reply.setResponse("response");

        std::string subscriptionId("subscriptionTestId");
        subscriptionPublication.setSubscriptionId(subscriptionId);
        subscriptionPublication.setResponse("publication");
    }

    void checkHeaderCreatorFromTo(const JoynrMessage& joynrMessage)
    {
        EXPECT_TRUE(joynrMessage.containsHeaderCreatorUserId());
        EXPECT_STREQ(senderID.c_str(), joynrMessage.getHeaderFrom().c_str());
        EXPECT_STREQ(receiverID.c_str(), joynrMessage.getHeaderTo().c_str());
        EXPECT_TRUE(joynrMessage.containsCustomHeader("test-header"));
        EXPECT_STREQ("test-header-value", joynrMessage.getCustomHeader("test-header").c_str());
    }

    void checkRequest(const JoynrMessage& joynrMessage)
    {
        // TODO create expected string from params and methodName
        std::stringstream expectedPayloadStream;
        expectedPayloadStream << R"({"_typeName":"joynr.Request",)";
        expectedPayloadStream << R"("methodName":"methodName",)";
        expectedPayloadStream << R"("paramDatatypes":["java.lang.Integer","java.lang.String"],)";
        expectedPayloadStream << R"("params":[42,"value"],)";
        expectedPayloadStream << R"("requestReplyId":")" << request.getRequestReplyId() << R"("})";
        std::string expectedPayload = expectedPayloadStream.str();
        EXPECT_EQ(expectedPayload, joynrMessage.getPayload());
    }

    void checkOneWayRequest(const JoynrMessage& joynrMessage)
    {
        // TODO create expected string from params and methodName
        std::stringstream expectedPayloadStream;
        expectedPayloadStream << R"({"_typeName":"joynr.OneWayRequest",)";
        expectedPayloadStream << R"("methodName":"methodName",)";
        expectedPayloadStream << R"("paramDatatypes":["java.lang.Integer","java.lang.String"],)";
        expectedPayloadStream << R"("params":[42,"value"])";
        expectedPayloadStream << R"(})";
        std::string expectedPayload = expectedPayloadStream.str();
        EXPECT_EQ(expectedPayload, joynrMessage.getPayload());
    }

    void checkReply(const JoynrMessage& joynrMessage)
    {
        std::stringstream expectedPayloadStream;
        expectedPayloadStream << R"({"_typeName":"joynr.Reply",)";
        expectedPayloadStream << R"("response":["response"],)";
        expectedPayloadStream << R"("requestReplyId":")" << reply.getRequestReplyId() << R"("})";
        std::string expectedPayload = expectedPayloadStream.str();
        EXPECT_EQ(expectedPayload, joynrMessage.getPayload());
    }

    void checkSubscriptionPublication(const JoynrMessage& joynrMessage)
    {
        std::stringstream expectedPayloadStream;
        expectedPayloadStream << R"({"_typeName":"joynr.SubscriptionPublication",)";
        expectedPayloadStream << R"("response":["publication"],)";
        expectedPayloadStream << R"("subscriptionId":")"
                              << subscriptionPublication.getSubscriptionId() << R"("})";
        std::string expectedPayload = expectedPayloadStream.str();
        EXPECT_EQ(expectedPayload, joynrMessage.getPayload());
    }

    void checkMulticastPublication(const JoynrMessage& joynrMessage,
                                   const MulticastPublication& multicastPublication)
    {
        std::stringstream expectedPayloadStream;
        expectedPayloadStream << R"({"_typeName":"joynr.MulticastPublication",)";
        expectedPayloadStream << R"("response":["publication"],)";
        expectedPayloadStream << R"("multicastId":")"
                              << multicastPublication.getMulticastId() << R"("})";
        std::string expectedPayload = expectedPayloadStream.str();
        EXPECT_EQ(expectedPayload, joynrMessage.getPayload());
    }

protected:
    ADD_LOGGER(JoynrMessageFactoryTest);
    JoynrMessageFactory messageFactory;
    std::string senderID;
    std::string receiverID;
    std::string requestReplyID;
    MessagingQos qos;
    Request request;
    OneWayRequest oneWayRequest;
    Reply reply;
    SubscriptionPublication subscriptionPublication;
};

INIT_LOGGER(JoynrMessageFactoryTest);

TEST_F(JoynrMessageFactoryTest, createRequest_withContentType)
{
    JoynrMessage joynrMessage = messageFactory.createRequest(senderID, receiverID, qos, request);
    checkHeaderCreatorFromTo(joynrMessage);
}

TEST_F(JoynrMessageFactoryTest, createRequest)
{
    JoynrMessage joynrMessage = messageFactory.createRequest(senderID, receiverID, qos, request);
    // warning if prepareRequest needs to long this assert will fail as it compares absolute
    // timestamps
    JoynrTimePoint now = std::chrono::time_point_cast<std::chrono::milliseconds>(
            std::chrono::system_clock::now());
    JoynrTimePoint expectedExpiryDate = now + std::chrono::milliseconds(qos.getTtl());
    JoynrTimePoint expiryDate = joynrMessage.getHeaderExpiryDate();
    EXPECT_NEAR(expectedExpiryDate.time_since_epoch().count(),
                expiryDate.time_since_epoch().count(),
                100.);
    JOYNR_LOG_DEBUG(logger,
                    "expiryDate: {} [{}]",
                    DispatcherUtils::convertAbsoluteTimeToTtlString(expiryDate),
                    std::chrono::duration_cast<std::chrono::milliseconds>(
                            expiryDate.time_since_epoch()).count());
    JOYNR_LOG_DEBUG(logger,
                    "expectedExpiryDate: {}  [{}]",
                    DispatcherUtils::convertAbsoluteTimeToTtlString(expectedExpiryDate),
                    std::chrono::duration_cast<std::chrono::milliseconds>(
                            expectedExpiryDate.time_since_epoch()).count());

    checkHeaderCreatorFromTo(joynrMessage);
    checkRequest(joynrMessage);
    EXPECT_EQ(JoynrMessage::VALUE_MESSAGE_TYPE_REQUEST, joynrMessage.getType());
}

TEST_F(JoynrMessageFactoryTest, createReply)
{
    JoynrMessage joynrMessage = messageFactory.createReply(senderID, receiverID, qos, reply);
    checkHeaderCreatorFromTo(joynrMessage);
    checkReply(joynrMessage);
    EXPECT_EQ(JoynrMessage::VALUE_MESSAGE_TYPE_REPLY, joynrMessage.getType());
}

TEST_F(JoynrMessageFactoryTest, createOneWayRequest)
{
    JoynrMessage joynrMessage =
            messageFactory.createOneWayRequest(senderID, receiverID, qos, oneWayRequest);
    checkHeaderCreatorFromTo(joynrMessage);
    checkOneWayRequest(joynrMessage);
    EXPECT_EQ(JoynrMessage::VALUE_MESSAGE_TYPE_ONE_WAY, joynrMessage.getType());
}

// TEST_F(JoynrMessageFactoryTest, createSubscriptionReply){
//    std::string subscriptionId("subscriptionTestId");
//    JoynrMessage joynrMessage = JoynrMessageFactory::prepareSubscriptionReply(senderID,
//    receiverID, payload, subscriptionId);
//    checkHeaderCreatorFromTo(joynrMessage);
//    checkPayload(joynrMessage);
//    EXPECT_QSTREQ(subscriptionId,
//    joynrMessage.getHeader<std::string>(JoynrMessage::HEADER_NAME_SUBSCRIPTION_ID));
//    EXPECT_QSTREQ(JoynrMessage::MESSAGE_TYPE_SUBSCRIPTION_REPLY, joynrMessage.getType());
//}

TEST_F(JoynrMessageFactoryTest, createPublication)
{
    JoynrMessage joynrMessage = messageFactory.createSubscriptionPublication(
            senderID, receiverID, qos, subscriptionPublication);
    checkHeaderCreatorFromTo(joynrMessage);
    checkSubscriptionPublication(joynrMessage);
    EXPECT_EQ(JoynrMessage::VALUE_MESSAGE_TYPE_PUBLICATION, joynrMessage.getType());
}

TEST_F(JoynrMessageFactoryTest, createMulticastPublication)
{
    MulticastPublication multicastPublication;
    multicastPublication.setMulticastId(receiverID);
    multicastPublication.setResponse("publication");
    JoynrMessage joynrMessage = messageFactory.createMulticastPublication(
            senderID, qos, multicastPublication);
    checkHeaderCreatorFromTo(joynrMessage);
    checkMulticastPublication(joynrMessage, multicastPublication);
    EXPECT_EQ(JoynrMessage::VALUE_MESSAGE_TYPE_MULTICAST, joynrMessage.getType());
}

TEST_F(JoynrMessageFactoryTest, createSubscriptionRequest)
{
    auto subscriptionQos = std::make_shared<OnChangeSubscriptionQos>();
    SubscriptionRequest subscriptionRequest;
    subscriptionRequest.setSubscriptionId("subscriptionId");
    subscriptionRequest.setSubscribeToName("attributeName");
    subscriptionRequest.setQos(subscriptionQos);
    JoynrMessage joynrMessage = messageFactory.createSubscriptionRequest(
            senderID, receiverID, qos, subscriptionRequest);
    checkHeaderCreatorFromTo(joynrMessage);
    EXPECT_EQ(JoynrMessage::VALUE_MESSAGE_TYPE_SUBSCRIPTION_REQUEST, joynrMessage.getType());
}

TEST_F(JoynrMessageFactoryTest, createMulticastSubscriptionRequest)
{
    auto subscriptionQos = std::make_shared<MulticastSubscriptionQos>();
    MulticastSubscriptionRequest subscriptionRequest;
    subscriptionRequest.setSubscriptionId("subscriptionId");
    subscriptionRequest.setMulticastId("multicastId");
    subscriptionRequest.setSubscribeToName("attributeName");
    subscriptionRequest.setQos(subscriptionQos);
    JoynrMessage joynrMessage = messageFactory.createMulticastSubscriptionRequest(
            senderID, receiverID, qos, subscriptionRequest);
    checkHeaderCreatorFromTo(joynrMessage);
    EXPECT_EQ(JoynrMessage::VALUE_MESSAGE_TYPE_MULTICAST_SUBSCRIPTION_REQUEST, joynrMessage.getType());
}

TEST_F(JoynrMessageFactoryTest, createSubscriptionStop)
{
    std::string subscriptionId("TEST-SubscriptionId");
    SubscriptionStop subscriptionStop;
    subscriptionStop.setSubscriptionId(subscriptionId);
    JoynrMessage joynrMessage =
            messageFactory.createSubscriptionStop(senderID, receiverID, qos, subscriptionStop);
    checkHeaderCreatorFromTo(joynrMessage);
    EXPECT_EQ(JoynrMessage::VALUE_MESSAGE_TYPE_SUBSCRIPTION_STOP, joynrMessage.getType());
}

TEST_F(JoynrMessageFactoryTest, testRequestContentType)
{
    Request request;
    request.setMethodName("methodName");
    request.setParams(std::string("test"));

    JoynrMessage message = messageFactory.createRequest(senderID, receiverID, qos, request);
    EXPECT_EQ(JoynrMessage::VALUE_CONTENT_TYPE_APPLICATION_JSON, message.getHeaderContentType());
}

TEST_F(JoynrMessageFactoryTest, testSetNoEffortHeader)
{
    JoynrMessage message = messageFactory.createRequest(senderID, receiverID, qos, request);
    EXPECT_FALSE(message.containsHeaderEffort());
}

TEST_F(JoynrMessageFactoryTest, testSetBestEffortHeader)
{
    qos.setEffort(MessagingQosEffort::Enum::BEST_EFFORT);
    JoynrMessage message = messageFactory.createRequest(senderID, receiverID, qos, request);
    EXPECT_TRUE(message.containsHeaderEffort());
    EXPECT_EQ(MessagingQosEffort::getLiteral(MessagingQosEffort::Enum::BEST_EFFORT),
              message.getHeaderEffort());
}

File: java/core/libjoynr/src/test/java/io/joynr/dispatching/JoynrMessageFactoryTest.java
package io.joynr.dispatching;

/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2016 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.mock;

import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Properties;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.google.inject.AbstractModule;
import com.google.inject.Guice;
import com.google.inject.Injector;
import com.google.inject.TypeLiteral;
import com.google.inject.multibindings.Multibinder;
import com.google.inject.name.Names;

import io.joynr.common.ExpiryDate;
import io.joynr.common.JoynrPropertiesModule;
import io.joynr.messaging.ConfigurableMessagingSettings;
import io.joynr.messaging.JsonMessageSerializerModule;
import io.joynr.messaging.MessagingQos;
import io.joynr.messaging.MessagingQosEffort;
import io.joynr.pubsub.SubscriptionQos;
import joynr.JoynrMessage;
import joynr.MulticastPublication;
import joynr.MulticastSubscriptionRequest;
import joynr.PeriodicSubscriptionQos;
import joynr.Reply;
import joynr.Request;
import joynr.SubscriptionPublication;
import joynr.SubscriptionRequest;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.runners.MockitoJUnitRunner;

@RunWith(MockitoJUnitRunner.class)
public class JoynrMessageFactoryTest {
    private static final long TTL = 1000;
    private static final long MAX_ALLOWED_EXPIRY_DATE_DIFF_MS = 500;
    private static final long NO_TTL_UPLIFT = 0;
    JoynrMessageFactory joynrMessageFactory;
    private String fromParticipantId;
    private String toParticipantId;
    private Request request;
    private Reply reply;
    private String payload;
    private ExpiryDate expiryDate;
    private MessagingQos messagingQos;
    private SubscriptionRequest subscriptionRequest;

    private SubscriptionPublication publication;
    private ObjectMapper objectMapper;

    @Before
    public void setUp() throws NoSuchMethodException, SecurityException {
        fromParticipantId = "sender";
        toParticipantId = "receiver";

        Injector injector = Guice.createInjector(new JoynrPropertiesModule(new Properties()),
                                                 new JsonMessageSerializerModule(),
                                                 new AbstractModule() {

                                                     @Override
                                                     protected void configure() {
                                                         bind(Long.class).annotatedWith(Names.named(ConfigurableMessagingSettings.PROPERTY_TTL_UPLIFT_MS))
                                                                         .toInstance(NO_TTL_UPLIFT);
                                                         requestStaticInjection(Request.class);
                                                         Multibinder<JoynrMessageProcessor> joynrMessageProcessorMultibinder = Multibinder.newSetBinder(binder(),
                                                                                                                                                        new TypeLiteral<JoynrMessageProcessor>() {
                                                                                                                                                        });
                                                         joynrMessageProcessorMultibinder.addBinding()
                                                                                         .toInstance(new JoynrMessageProcessor() {
                                                                                             @Override
                                                                                             public JoynrMessage process(JoynrMessage joynrMessage) {
                                                                                                 joynrMessage.getHeader()
                                                                                                             .put("test",
                                                                                                                  "test");
                                                                                                 return joynrMessage;
                                                                                             }
                                                                                         });
                                                     }

                                                 });

        objectMapper = injector.getInstance(ObjectMapper.class);

        payload = "payload";
        Method method = TestRequestCaller.class.getMethod("respond", new Class[]{ String.class });
        request = new Request(method.getName(), new String[]{ payload }, method.getParameterTypes());
        String requestReplyId = request.getRequestReplyId();
        reply = new Reply(requestReplyId, objectMapper.<JsonNode> valueToTree(payload));
        messagingQos = new MessagingQos(TTL);
        expiryDate = DispatcherUtils.convertTtlToExpirationDate(messagingQos.getRoundTripTtl_ms());

        String subscriptionId = "subscription";
        String attributeName = "attribute";
        PeriodicSubscriptionQos subscriptionqos = new PeriodicSubscriptionQos();
        subscriptionqos.setPeriodMs(1000).setValidityMs(10).setAlertAfterIntervalMs(1500).setPublicationTtlMs(1000);
        subscriptionRequest = new SubscriptionRequest(subscriptionId, attributeName, subscriptionqos);
        String response = "response";
        publication = new SubscriptionPublication(Arrays.asList(response), subscriptionId);

        joynrMessageFactory = injector.getInstance(JoynrMessageFactory.class);
    }

    public static void assertExpiryDateEquals(long expectedValue, JoynrMessage message) {
        String headerExpiryString = message.getHeaderValue(JoynrMessage.HEADER_NAME_EXPIRY_DATE);
        long headerExpiryValue = Long.parseLong(headerExpiryString);
        long diff = Math.abs(expectedValue - headerExpiryValue);
        assertTrue("ExpiryDate=" + headerExpiryString + " differs " + diff + "ms (more than "
                           + MAX_ALLOWED_EXPIRY_DATE_DIFF_MS + "ms) from the expected value=" + expectedValue,
                   (diff <= MAX_ALLOWED_EXPIRY_DATE_DIFF_MS));
    }

    @Test
    public void createRequest() {
        JoynrMessage message = joynrMessageFactory.createRequest(fromParticipantId,
                                                                 toParticipantId,
                                                                 request,
                                                                 messagingQos);
        assertEquals(JoynrMessage.MESSAGE_TYPE_REQUEST, message.getType());

        assertEquals(fromParticipantId, message.getHeaderValue(JoynrMessage.HEADER_NAME_FROM_PARTICIPANT_ID));
        assertEquals(toParticipantId, message.getHeaderValue(JoynrMessage.HEADER_NAME_TO_PARTICIPANT_ID));

        assertExpiryDateEquals(expiryDate.getValue(), message);
        assertTrue(message.getPayload() != null);
        assertNotNull(message.getCreatorUserId());
    }

    @Test
    public void createRequestWithCustomEffort() {
        MessagingQos customMessagingQos = new MessagingQos();
        customMessagingQos.setEffort(MessagingQosEffort.BEST_EFFORT);
        JoynrMessage message = joynrMessageFactory.createRequest(fromParticipantId,
                                                                 toParticipantId,
                                                                 request,
                                                                 customMessagingQos);
        expiryDate = DispatcherUtils.convertTtlToExpirationDate(customMessagingQos.getRoundTripTtl_ms());
        assertExpiryDateEquals(expiryDate.getValue(), message);
        assertEquals(String.valueOf(MessagingQosEffort.BEST_EFFORT),
                     message.getHeaderValue(JoynrMessage.HEADER_NAME_EFFORT));
    }

    @Test
    public void createRequestWithCustomHeaders() {
        final Map<String, String> myCustomHeaders = new HashMap<>();
        final String headerName = "header";
        final String headerValue = "value";
        myCustomHeaders.put(headerName, headerValue);
        messagingQos.getCustomMessageHeaders().putAll(myCustomHeaders);
        JoynrMessage message = joynrMessageFactory.createRequest(fromParticipantId,
                                                                 toParticipantId,
                                                                 request,
                                                                 messagingQos);
        assertEquals(JoynrMessage.MESSAGE_TYPE_REQUEST, message.getType());
        assertExpiryDateEquals(expiryDate.getValue(), message);

        final String expectedCustomHeaderName = JoynrMessage.MESSAGE_CUSTOM_HEADER_PREFIX + headerName;
        assertTrue(message.getHeader().containsKey(expectedCustomHeaderName));
        Map<String, String> customHeaders = message.getCustomHeaders();
        assertTrue(customHeaders.size() == 1);
        assertTrue(customHeaders.containsKey(headerName));
    }

    @Test
    public void testCreateOneWayRequest() {
        JoynrMessage joynrMessage = joynrMessageFactory.createOneWayRequest(fromParticipantId,
                                                                            toParticipantId,
                                                                            request,
                                                                            messagingQos);
        assertNotNull(joynrMessage);
        assertEquals(JoynrMessage.MESSAGE_TYPE_ONE_WAY, joynrMessage.getType());
        assertEquals(fromParticipantId, joynrMessage.getHeaderValue(JoynrMessage.HEADER_NAME_FROM_PARTICIPANT_ID));
        assertEquals(toParticipantId, joynrMessage.getHeaderValue(JoynrMessage.HEADER_NAME_TO_PARTICIPANT_ID));
        assertExpiryDateEquals(expiryDate.getValue(), joynrMessage);
        assertNotNull(joynrMessage.getPayload());
        assertNotNull(joynrMessage.getCreatorUserId());
    }

    @Test
    public void testCreateOneWayRequestWithCustomHeader() {
        final Map<String, String> myCustomHeaders = new HashMap<>();
        final String headerName = "header";
        final String headerValue = "value";
        myCustomHeaders.put(headerName, headerValue);
        messagingQos.getCustomMessageHeaders().putAll(myCustomHeaders);
        JoynrMessage message = joynrMessageFactory.createOneWayRequest(fromParticipantId,
                                                                            toParticipantId,
                                                                            request,
                                                                            messagingQos);
        assertNotNull(message);
        assertEquals(JoynrMessage.MESSAGE_TYPE_ONE_WAY, message.getType());
        assertExpiryDateEquals(expiryDate.getValue(), message);
        final String expectedCustomHeaderName = JoynrMessage.MESSAGE_CUSTOM_HEADER_PREFIX + headerName;
        assertTrue(message.getHeader().containsKey(expectedCustomHeaderName));
        Map<String, String> customHeaders = message.getCustomHeaders();
        assertTrue(customHeaders.size() == 1);
        assertTrue(customHeaders.containsKey(headerName));

    }

    @Test
    public void createReply() {
        JoynrMessage message = joynrMessageFactory.createReply(fromParticipantId, toParticipantId, reply, messagingQos);

        assertEquals(JoynrMessage.MESSAGE_TYPE_REPLY, message.getType());
        assertEquals(fromParticipantId, message.getHeaderValue(JoynrMessage.HEADER_NAME_FROM_PARTICIPANT_ID));
        assertEquals(toParticipantId, message.getHeaderValue(JoynrMessage.HEADER_NAME_TO_PARTICIPANT_ID));
        assertEquals(JoynrMessage.CONTENT_TYPE_APPLICATION_JSON,
                     message.getHeaderValue(JoynrMessage.HEADER_NAME_CONTENT_TYPE));
        assertExpiryDateEquals(expiryDate.getValue(), message);

        assertTrue(message.getPayload() != null);
        assertNotNull(message.getCreatorUserId());
    }

    @Test
    public void createSubscriptionRequest() {
        JoynrMessage message = joynrMessageFactory.createSubscriptionRequest(fromParticipantId,
                                                                             toParticipantId,
                                                                             subscriptionRequest,
                                                                             messagingQos);
        assertEquals(JoynrMessage.MESSAGE_TYPE_SUBSCRIPTION_REQUEST, message.getType());
        assertEquals(fromParticipantId, message.getHeaderValue(JoynrMessage.HEADER_NAME_FROM_PARTICIPANT_ID));
        assertEquals(toParticipantId, message.getHeaderValue(JoynrMessage.HEADER_NAME_TO_PARTICIPANT_ID));
        assertExpiryDateEquals(expiryDate.getValue(), message);

        assertTrue(message.getPayload() != null);
        assertNotNull(message.getCreatorUserId());
    }

    @Test
    public void createPublication() {
        JoynrMessage message = joynrMessageFactory.createPublication(fromParticipantId,
                                                                     toParticipantId,
                                                                     publication,
                                                                     messagingQos);
        assertEquals(JoynrMessage.MESSAGE_TYPE_PUBLICATION, message.getType());
        assertEquals(fromParticipantId, message.getHeaderValue(JoynrMessage.HEADER_NAME_FROM_PARTICIPANT_ID));
        assertEquals(toParticipantId, message.getHeaderValue(JoynrMessage.HEADER_NAME_TO_PARTICIPANT_ID));
        assertExpiryDateEquals(expiryDate.getValue(), message);

        assertTrue(message.getPayload() != null);
        assertNotNull(message.getCreatorUserId());
    }

    @Test
    public void testMessageProcessorUsed() {
        JoynrMessage joynrMessage = joynrMessageFactory.createRequest("from",
                                                                      "to",
                                                                      new Request("name", new Object[0], new Class[0]),
                                                                      new MessagingQos());
        assertNotNull(joynrMessage.getHeader().get("test"));
        assertEquals("test", joynrMessage.getHeader().get("test"));
    }

    @Test
    public void testCreateMulticastMessage() {
        String multicastId = "multicastId";
        MulticastPublication multicastPublication = new MulticastPublication(Collections.emptyList(), multicastId);

        JoynrMessage joynrMessage = joynrMessageFactory.createMulticast(fromParticipantId,
                                                                        multicastPublication,
                                                                        messagingQos);

        assertNotNull(joynrMessage);
        assertExpiryDateEquals(expiryDate.getValue(), joynrMessage);
        assertEquals(fromParticipantId, joynrMessage.getFrom());
        assertEquals(multicastId, joynrMessage.getTo());
        assertEquals(JoynrMessage.MESSAGE_TYPE_MULTICAST, joynrMessage.getType());
        assertTrue(joynrMessage.getPayload().contains(MulticastPublication.class.getName()));
    }

    @Test
    public void testCreateMulticastSubscriptionRequest() {
        String multicastId = "multicastId";
        String subscriptionId = "subscriptionId";
        String multicastName = "multicastName";
        SubscriptionQos subscriptionQos = mock(SubscriptionQos.class);

        MulticastSubscriptionRequest multicastSubscriptionRequest = new MulticastSubscriptionRequest(multicastId,
                                                                                                     subscriptionId,
                                                                                                     multicastName,
                                                                                                     subscriptionQos);

        JoynrMessage result = joynrMessageFactory.createSubscriptionRequest(fromParticipantId,
                                                                            toParticipantId,
                                                                            multicastSubscriptionRequest,
                                                                            messagingQos);

        assertNotNull(result);
        assertEquals(JoynrMessage.MESSAGE_TYPE_MULTICAST_SUBSCRIPTION_REQUEST, result.getType());
        assertExpiryDateEquals(expiryDate.getValue(), result);
    }

}
